src/components/ingestion/AbstractIngestionEngine.js

```diff
@@ -1,66 +1,120 @@
 import { spawn } from 'child_process'
+import {
+	pipeline,
+	Writable,
+	Transform,
+} from 'stream'
 import { Payload } from '@tvkitchen/base-classes'
 import { dataTypes } from '@tvkitchen/base-constants'
 import {
 	AbstractInstantiationError,
 	NotImplementedError,
 } from '@tvkitchen/base-errors'
+import { TSDemuxer } from 'ts-demuxer'
+
 import logger from '%src/lib/logger'
 import kafka from '%src/lib/kafka'
+import { tsToSeconds } from '%src/lib/utils/mpegts'
 
 class AbstractIngestionEngine {
 	constructor() {
 		if (this.constructor === AbstractIngestionEngine) {
 			throw new AbstractInstantiationError(this.constructor.name)
 		}
 		this.producer = kafka.producer()
+		this.mpegtsDemuxer = new TSDemuxer(this.processDemuxedPacket)
+		this.mostRecentDemuxedPacket = null
 	}
 
 	/**
-	 * Ingests data into the TV Kitchen pipeline.
+	 * Ingests an mpegts payload into the TV Kitchen pipeline.
 	 *
-	 * @param  {Buffer} data The buffer of data to be ingested.
-	 * @return {Boolean}     Whether or not the buffer was successfully ingested.
+	 * @param  {Payload} data The payload to be ingested.
+	 * @return {Boolean}      Whether or not the payload was successfully ingested.
 	 */
-	ingestData = async (data) => {
-		const payload = new Payload()
-		payload.data = data
-		await this.producer.send({
-			topic: dataTypes.STREAM.DATA,
-			messages: [{
-				value: data,
-			}],
-		})
-		return true
+	ingestPayloadStream = Writable({
+		objectMode: true,
+		write: async (payload, enc, done) => {
+			if (!(payload instanceof Payload)) {
+				done(new Error('ingestPayloadStream received non-Payload data'))
+			} else {
+				done(this.producer.send({
+					topic: dataTypes.STREAM.CONTAINER,
+					messages: [{
+						value: Payload.serialize(payload),
+					}],
+				}))
+			}
+		},
+	})
+
+	/**
+	 * Updates ingestion state based on the latest demuxed packet data.
+	 *
+	 * This method is called by our mpegts demuxer, and allows the ingestion engine to track
+	 * the most recent demuxed packet.
+	 *
+	 * @param  {Packet} packet The latest TSDemuxer Packet object.
+	 */
+	processDemuxedPacket = (packet) => {
```

Comment:
😬 Okay so feel free to discard this because we're bike-shedding now _but_ this isn't really _processing_ the packet, is it? I mean it's the callback that `TSDemuxer` calls when _it_ has processed the packet, and handed it back to you to do whatever with. So I feel like it should really be called something like `onPacketDemuxingComplete` or something? That would certainly help _me_ understand it with less cartography.

Reply:
I saw this as the ingestion engine processing a demuxed packet -- like, not to generate it but more like:

~TSDemuxer Delivery Person Enters Scene~

*Delivery Person (smiling pleasantly):*  Hello there!  Here is `A DEMUXED PACKET`!  Hope you're having a great day.

~Delivery Person proudly places the package on the desk~

*AbstractIngestionEngine Employee (bored):* thanks.  I'll process it now.

~AbstractIngestionEngine Employee takes the packet and proceeds to apathetically slide it into the trash can labeled `MOST RECENT PACKET`~

Reply:
Open to swapping `process` with something like `handle` (if maybe `process` evokes the concept of `extract from a deeper stream`)

`ingest` could work as well.

Reply:
Bless you for using narrative dialogue, and explanation format I understand! (That… genuinely did help me to think through it properly.)

I think you can safely leave it `process` given your explanation, though I'd probably err on the side of `handle` or `on` myself. (`complete` only worked in my original one because it was the _packet_ demuxing, not stream demuxing, that was complete… but I agree it is a distractingly final word.)

Anyway, go forth!

Reply:
Changing the name to `onDemuxedPacket`

I also added a big docblock for the class itself to describe the overall pipeline (though I do not use narrative dialogue unfortunately)